src/hotspot/share/memory/metaspace.cpp

```diff
@@ -581,86 +582,78 @@ bool Metaspace::class_space_is_initialized() {
   return MetaspaceContext::context_class() != nullptr;
 }
 
-// Reserve a range of memory at an address suitable for en/decoding narrow
-// Klass pointers (see: CompressedClassPointers::is_valid_base()).
-// The returned address shall both be suitable as a compressed class pointers
-//  base, and aligned to Metaspace::reserve_alignment (which is equal to or a
-//  multiple of allocation granularity).
-// On error, returns an unreserved space.
-ReservedSpace Metaspace::reserve_address_space_for_compressed_classes(size_t size) {
-
-#if defined(AARCH64) || defined(PPC64)
-  const size_t alignment = Metaspace::reserve_alignment();
-
-  // AArch64: Try to align metaspace class space so that we can decode a
-  // compressed klass with a single MOVK instruction. We can do this iff the
-  // compressed class base is a multiple of 4G.
-  // Additionally, above 32G, ensure the lower LogKlassAlignmentInBytes bits
-  // of the upper 32-bits of the address are zero so we can handle a shift
-  // when decoding.
-
-  // PPC64: smaller heaps up to 2g will be mapped just below 4g. Then the
-  // attempt to place the compressed class space just after the heap fails on
-  // Linux 4.1.42 and higher because the launcher is loaded at 4g
-  // (ELF_ET_DYN_BASE). In that case we reach here and search the address space
-  // below 32g to get a zerobased CCS. For simplicity we reuse the search
-  // strategy for AARCH64.
-
-  static const struct {
-    address from;
-    address to;
-    size_t increment;
-  } search_ranges[] = {
-    {  (address)(4*G),   (address)(32*G),   4*G, },
-    {  (address)(32*G),  (address)(1024*G), (4 << LogKlassAlignmentInBytes) * G },
-    {  nullptr, nullptr, 0 }
-  };
-
-  // Calculate a list of all possible values for the starting address for the
-  // compressed class space.
-  ResourceMark rm;
-  GrowableArray<address> list(36);
-  for (int i = 0; search_ranges[i].from != nullptr; i ++) {
-    address a = search_ranges[i].from;
-    assert(CompressedKlassPointers::is_valid_base(a), "Sanity");
-    while (a < search_ranges[i].to) {
-      list.append(a);
-      a +=  search_ranges[i].increment;
+// Reserve a range of memory that is to contain narrow Klass IDs. If "cds_runtime"
+// is true, the memory must be located at an address that can be directly used as
+// encoding base, otherwise the API has more leeway (e.g. optimizing reservation
+// for zero-based encoding).
+ReservedSpace Metaspace::reserve_address_space_for_compressed_classes(size_t size, bool cds_runtime) {
+
+  char* result = nullptr;
+  const bool randomize = RandomizeClassSpaceLocation;
+
+  // At CDS runtime, there is no need bothering with low-address reservation since we cannot set
+  // the encoding base to zero anyway: the encoding base has to be the base of the mapped archive
+  // (ultimately, the start address of the region we are reserving here).
+  const bool try_in_low_address_ranges = !cds_runtime;
```

Comment:
I think `try_in_low_address_range` would be a far better name for the method param than either the current `cds_runtime` or your original `strict_base`, following the principle that *it does what it says on the tin*.

Reply:
Bike shedding done, you may stick with the current name if you wish.

Reply:
I like `try_in_low_address_range`, I will use that. Note that we hopefully can remove this clutch in the future.